PLAT-47: Add authentication to the login component

dondrzzy opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111 Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber JiraIssueTitle' -->
## JIRA TICKET NAME
[PLAT-47: Add authentication to the login component](https://issues.openmrs.org/browse/PLAT-47)

## Description of what I changed

- Added authentication to the login component.
- Authorized users are redirected to "/" where the application is loaded
- Show error messages for unauthorized users
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

## Issue I worked on
[PLAT-47](https://issues.openmrs.org/browse/PLAT-47)
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

dkayiwa commented:
@dondrzzy can you take a look at this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

dondrzzy commented:
@dkayiwa, I have looked at it. I have added the compiled application to the ticket on the Jira board. I have also added a screen recording of the procedure/authentication.

emmabaye reviewed (approved):
LGTM

IAMOTZ commented on app/js/components/Login.jsx:
```diff
@@ -6,37 +6,73 @@
+import React, { Component } from 'react';
+import { ToastContainer, toast } from 'react-toastify';
+import axios from 'axios';
+import axiosInstance from '../config/axiosInstance';
```
Since you are using the `axionInstance` to make your API call, you don't need to import `axios` again. You should remove this line:
```
import axios from 'axios';
```

dondrzzy replied:
Oh, I accidentally forgot that there. Thanks. Addressed

IAMOTZ commented on app/js/components/Login.jsx:
```diff
@@ -6,37 +6,73 @@
     const { username, password } = this.state;
-    this.props.history.push("/")
+    const auth = 'Basic '+ new Buffer(`${username}:${password}`).toString('base64');
+    this.authenticateUser(`/session`, auth)
```
Can we really trust the `username` and `password` input from the User, how about you add some user input validations before sending the info to the server?

dondrzzy replied:
I think having to do regex matching for valid usernames and strong passwords on login is rarely done. To sign up, I totally agree. Secondly, limiting usernames to a certain pattern on the front end must always be in sync with what is on the backend in case someone disables javascript so as to show the same errors. I am not certain what conventions are used in the Java backend for what passes as a valid username. Password validation on login seems like a stretch too.
Take a look at 
- https://web.facebook.com/login.php
- Gmail signIn
- https://www.amazon.com/ap/signin
Validation is hardly done on login at the client side.

dondrzzy replied:
Definitely not. I agree with you.  I have added empty form field validation in case a user edits the HTML:5 required attribute. Usernames take on many forms unless agreed upon and passwords are not validated on login.
- https://www.amazon.com/ap/signin
- Gmail sign in
- https://web.facebook.com/login.php
I guess it is because the user input is not stored but used in comparison with what was stored at sign up.

IAMOTZ replied:
Nice one.

alexmochu commented:
LGTM

IAMOTZ commented on __tests__/Login.test.js:
```diff
@@ -48,32 +47,107 @@ describe('Login Component', () => {
+    expect(spy).toHaveBeenCalledWith(
+      'Oops, something went wrong. Contact the system adminstrator'
+    );
+  });
```
Please add some order to your tests to make it easier to read and understand.
You can group the tests for `'handleError`, `handleResponse`, `handleChange`, `'authenticateUser ` methods in different `describe` blocks.

dondrzzy replied:
Alright, on it. It will make the tests more readable.

dondrzzy replied:
Addressed

dotNesh reviewed (approved):
LGTM

IAMOTZ reviewed (approved):
Job Well Done @dondrzzy.
Looks Good To Me.

Commit:
PLAT-47: Add authentication to the login component
